core/matrix/diagonal.cpp

```diff
@@ -67,20 +68,21 @@ void Diagonal<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
 {
     auto exec = this->get_executor();
 
-    if (dynamic_cast<const Dense<ValueType> *>(b) &&
-        dynamic_cast<Dense<ValueType> *>(x)) {
-        exec->run(diagonal::make_apply_to_dense(this, as<Dense<ValueType>>(b),
-                                                as<Dense<ValueType>>(x)));
-    } else if (dynamic_cast<const Csr<ValueType, int32> *>(b) &&
-               dynamic_cast<Csr<ValueType, int32> *>(x)) {
+    if (dynamic_cast<const Csr<ValueType, int32> *>(b) &&
+        dynamic_cast<Csr<ValueType, int32> *>(x)) {
         exec->run(diagonal::make_apply_to_csr(
             this, as<Csr<ValueType, int32>>(b), as<Csr<ValueType, int32>>(x)));
     } else if (dynamic_cast<const Csr<ValueType, int64> *>(b) &&
                dynamic_cast<Csr<ValueType, int64> *>(x)) {
         exec->run(diagonal::make_apply_to_csr(
             this, as<Csr<ValueType, int64>>(b), as<Csr<ValueType, int64>>(x)));
     } else {
-        GKO_NOT_IMPLEMENTED;
+        precision_dispatch_spmv<ValueType>(
+            [&](auto dense_b, auto dense_x) {
+                exec->run(
+                    diagonal::make_apply_to_dense(this, dense_b, dense_x));
+            },
+            b, x);
     }
```

Comment:
In terms of functionality that should work as all formats can convert to Dense, but do we want that to happen, what about other formats? We  could keep a `GKO_NOT_IMPLEMENTED` as well.

Reply:
I guess the new form corresponds more to what is in other cases like for CSR where we assume all other cases to be dense.

Reply:
I prefer staying `GKO_NOT_IMPLEMENTED`. Converting to dense may needs a lot of storage

Reply:
True, I hadn't even considered the sparse -> dense conversions here. So I will instead have to test against all Dense types instead.

Reply:
I agree, diagonal to dense might not be something one would want to do.

Reply:
I actually misremembered my implementation, I never try to cast to ConvertibleTo in make_temporary_conversion, only to Dense directly, so this is not an issue.

Reply:
What I mean here is that when someone passes a `Csr` type or any other form of LinOp instead of a `Dense`, before we would got an exception thrown here `GKO_NOT_IMPLEMENTED`. Now what would happen, in what way would it fail, it's actually not clear looking at the code of `make_temporary_conversion`.

Exceptions are part of the interface AFAIK?

Reply:
Before my last commit, it would have thrown NotSupported in make_temporary_conversion. Now with the new changes in place, it would throw NotSupported in conversion_helper:::convert, but now that I think of it, the previous version might actually be better, together with a `make_temporary_conversion/precision_dispatch_nothrow` that just returns nullptr or something like that in case of an error.